Aplib.Core/Desire/Goal.cs

```diff
@@ -0,0 +1,79 @@
+namespace Aplib.Core.Desire
+{
+    public class Goal
+    {
+        /// <summary>
+        /// The goal is considered to be completed, when the distance of the
+        /// <see cref="CurrentHeuristics"/> is below this value.
+        /// </summary>
+        private const float Epsilon = 0.005f;
```

Comment:
I agree with Joachim on his points. Also, I have some doubts regarding the current heuristic mechanism used. I am not sure what the current goal of the heuristic mechanism is compared to a simple goal predicate. I guess that a heuristic is used to either ...:

**... allow more flexibility in goal predicates (i.e., 'termination predicates', 'when has the goal been completed?').**
- Firstly, I think the above does not define a heuristic. If the above description is the purpose of a 'heuristic' in this code, the naming is a bit confusing to me and does not align with what I would expect it to mean.
- If the above is what is meant, then the same functionality can easily be achieved by a simple boolean predicate. On the contrary, a simple boolean predicate would become unnecessarily complicated.
- A boolean predicate, or even a guard, would probably make it significantly easier for the user to reason about his logic, compared to a heuristic. A few lectures can be devoted to explain what a heuristic is to _CS students_, so I don't think it would appeal to the average Unity developer.
- Long story short, if the above is the intended purpose, I would completely remove the concept of heuristic and replace it with a boolean predicate as I do not see any benefit of a heuristic if used like this.

**... decide on how to proceed in the goal structure (i.e., 'which goal to choose?')**
- One might want to use a different 'termination predicate' than a heuristic. Right now this is not possible (or at the very best considerably difficult to achieve).
- Same point as for the other interpretation of heuristic, but for completeness I will put it here again: a boolean predicate, or even a guard, would probably make it significantly easier for the user to reason about his logic, compared to a heuristic. A few lectures can be devoted to explain what a heuristic is to _CS students_, so I don't think it would appeal to the average Unity developer.
- Another point that applies here as well: the same functionality can easily be achieved by a simple boolean predicate. On the contrary, a simple boolean predicate would become unnecessarily complicated.
- Long story short, for the above purpose, I think the idea of a heuristic is a great idea! However, I think it should be optional, maybe even only part of a separate special Goal class. I would for now remove the concept of heuristic from the goal class, remove it with a boolean predicate, and add the idea of a heuristic for goal structure decisions to the backlog to be researched (I think it has great potential for extended functionality).

Reply:
Please look at this before resolving the conversation :).

Reply:
Yeah I accidentally resolved this issue, and could not find it back. Will ook into this tomorrow

Reply:
No problem, I should have put it in a different conversation